llvm/test/Analysis/CostModel/RISCV/cast.ll

```diff
@@ -1108,60 +1108,60 @@ define void @trunc() {
 ; RV32-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %nxv1i64_nxv1i1 = trunc <vscale x 1 x i64> undef to <vscale x 1 x i1>
 ; RV32-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %nxv2i16_nxv2i8 = trunc <vscale x 2 x i16> undef to <vscale x 2 x i8>
 ; RV32-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %nxv2i32_nxv2i8 = trunc <vscale x 2 x i32> undef to <vscale x 2 x i8>
-; RV32-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %nxv2i64_nxv2i8 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i8>
+; RV32-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %nxv2i64_nxv2i8 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i8>
 ; RV32-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %nxv2i32_nxv2i16 = trunc <vscale x 2 x i32> undef to <vscale x 2 x i16>
-; RV32-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %nxv2i64_nxv2i16 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i16>
-; RV32-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %nxv2i64_nxv2i32 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i32>
+; RV32-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %nxv2i64_nxv2i16 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i16>
+; RV32-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %nxv2i64_nxv2i32 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i32>
```

Comment:
In this case, (LMUL2, e64) is truncated to (LMUL1, e32).
`Cost Model: Found an estimated cost of 1 for instruction: %nxv2i64_nxv2i32 = trunc <vscale x 2 x i64> undef to <vscale x 2 x i32>`
I believe the cost is implementation-defined. It can be executed either as a single uop reading two registers from `vs2` or as two uops reading different parts from `vs2`. In the former case, the cost is 1, and in the latter case, it is 2.
The implementation-related decision would be encapsulated in `getRISCVInstructionCost(RISCV::VNSRL_WI)`, so we should always pass the destination type to `getRISCVInstructionCost`.

Reply:
In general, we model the cost of an LMUL operation proportional to the number of registers read or written - not just written.  As an example, consider that we give reductions a non-unit cost.  In that particular case, we actually scale by log(VL) so it's not a perfect analogy.  

As for the point about encapsulating this change in getRISCVInstructionCost, I'd be fine shifting to that approach.  On reflection, passing in the destination type (which matches the instruction semantic of defining it's output in terms of SEW and input as 2 x SEW) seems reasonable.  

If we do that, we probably want to model the other narrowing instructions (vnsra, vnclip, vfncvt), in an analogous manner.

Reply:
I went and ran a couple micro benchmarks on the bp3.

```
Running vnsrl-mf2.out
  ~3.934785 cycles-per-inst
  ~4053.832000 cycles-per-iteration
  ~1030.255150 insts-per-iteration

Running vnsrl-m1.out
  ~3.971956 cycles-per-inst
  ~4020.713650 cycles-per-iteration
  ~1012.275450 insts-per-iteration

Running vnsrl-m2.out
  ~7.453559 cycles-per-inst
  ~8139.421000 cycles-per-iteration
  ~1092.018050 insts-per-iteration

Running vnsrl-m4.out
  ~14.008862 cycles-per-inst
  ~16234.939100 cycles-per-iteration
  ~1158.904950 insts-per-iteration

For comparison, here's a vadd.vv at m1:
  ~3.970621 cycles-per-inst
  ~4017.622100 cycles-per-iteration
  ~1011.837100 insts-per-iteration
```
So, at least on this board, it looks like you're right that the cost is scaling with the destination LMUL, not the source LMUL.  Interesting!

Reply:
The throughput tests show that the processing time doubles when the LMUL doubles.
I am not sure if I understand it correctly, but it looks like the vnsrl on this machine is capable to output VLEN/4 per cycle.
Based on the observation that CPI is ~1 for mf4, ~2 for mf2, ~4 for m1, and ~8 for m2, regardless of sew.
And I would say the number of uops for each vnsrl is LMUL*4.
It is just assumption because I don't know the micro-architecture.
Based on my understanding, vnsrl on sifive-x280 is capable to handle DLEN/2 per cycle, and vnsrl on sifive-p670 is DLEN  per cycle.